src/js/letters/components/Address.jsx

```diff
@@ -0,0 +1,151 @@
+import React from 'react';
+import _ from 'lodash';
+import { set } from 'lodash/fp';
+
+import ErrorableSelect from './ErrorableSelect';
+import ErrorableTextInput from './ErrorableTextInput';
+import { isNotBlank, isBlankAddress, isValidUSZipCode, isValidCanPostalCode } from '../../common/utils/validations';
+import { countries, states } from '../../common/utils/options-for-select';
+
+/**
+ * Input component for an address.
+ *
+ * No validation is provided through a currently stubbed isAddressValid function.
+ */
+class Address extends React.Component {
+  constructor() {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.validateAddressField = this.validateAddressField.bind(this);
+    this.validatePostalCode = this.validatePostalCode.bind(this);
+  }
+
+  componentWillMount() {
+    this.id = _.uniqueId('address-input-');
+  }
+
+  // TODO: Look into if this is the best way to update address,
+  // it is incredibly slow right now
+  handleChange(path, update) {
+    let address = set(path, update, this.props.value);
+    // if country is changing we should clear the state
+    if (path === 'country') {
+      address = set('state', '', address);
+    }
+
+    this.props.onUserInput(address);
+  }
+
+  validateAddressField(field) {
+    if (this.props.required || !isBlankAddress(this.props.value)) {
+      return isNotBlank(field);
+    }
+
+    return true;
+  }
+
+  validatePostalCode(postalCodeField) {
+    let isValid = true;
+
+    if (this.props.required || !isBlankAddress(this.props.value)) {
+      isValid = isValid && isNotBlank(postalCodeField);
+    }
+
+    if (this.props.value.country === 'USA' && isNotBlank(postalCodeField)) {
+      isValid = isValid && isValidUSZipCode(postalCodeField);
+    }
+
+    if (this.props.value.country === 'CAN' && isNotBlank(postalCodeField)) {
+      isValid = isValid && isValidCanPostalCode(postalCodeField);
+    }
+
+    return isValid;
+  }
+
+  isMilitaryCity(city) {
+    const lowerCity = city.toLowerCase().trim();
+
+    return lowerCity === 'apo' || lowerCity === 'fpo' || lowerCity === 'dpo';
+  }
+
+  render() {
+    let stateList = [];
+    const selectedCountry = this.props.value.country === 'US' ? 'USA' : this.props.value.country;
+    if (states[selectedCountry]) {
+      stateList = states[selectedCountry];
+      if (this.props.value.city && this.isMilitaryCity(this.props.value.city)) {
+        stateList = stateList.filter(state => state.value === 'AE' || state.value === 'AP' || state.value === 'AA');
+      }
+    }
+
+    const stateProvince = _.hasIn(states, this.props.value.country === 'US' ? 'USA' : this.props.value.country)
```

Comment:
This part 
```
this.props.value.country === 'US' ? 'USA' : this.props.value.country
```
looks like it can be replaced with `selectedCounrty`.

I see this little bit of logic duplicated in a few places, maybe it belongs in some kind of helper function so we can re-use?

Reply:
This code has changed since you made the comment, but basically the reason why I had to do this logic of `this.props.value.country === 'US' ? 'USA' : this.props.value.country` in several places is because the address we get for the user under `destination` has the country listed as `US`, but in the list of countries we get from EVSS from the `v0/address/countries` endpoint, it is listed as `USA`. Because they don't match, in order for the country to be selected at US once the fields load I have to set the value as `USA`. This is a question I plan on asking EVSS about, if there's a reason why we're storing 2 different values for the U.S., so this may change.

Reply:
Sounds good.  I don't doubt the need for the logic, just that it's not DRY.

If you can get EVSS to clean up their API, then the need to this logic disappears altogether! 👍

Reply:
So a pretty easy change - line 81 could read:
```jsx
const stateProvince = _.hasIn(states, selectedCountry)
```

& line 104
```jsx
value={selectedCountry}
```

Reply:
I've changed this in a separate branch I have for #4520 that is branched off of this one. I really want to just merge this as is, and this will be fixed in a separate PR.